src/mongoc/mongoc-client-pool.c

```diff
@@ -193,23 +198,47 @@ mongoc_client_pool_push (mongoc_client_pool_t *pool,
 {
    ENTRY;
 
-   bson_return_if_fail(pool);
-   bson_return_if_fail(client);
+   bson_return_if_fail (pool);
+   bson_return_if_fail (client);
 
-   /*
-    * TODO: Shutdown old client connections.
-    */
+   if (pool->size > pool->min_pool_size) {
+      mongoc_client_t *old_client;
+      old_client = _mongoc_queue_pop_head (&pool->queue);
+      mongoc_client_destroy (old_client);
+      pool->size--;
+   }
 
-   /*
-    * TODO: We should try to make a client healthy again if it
-    *       is unhealthy since this is typically where a thread
-    *       is done with its work.
-    */
+   if ((client->cluster.state == MONGOC_CLUSTER_STATE_HEALTHY) ||
+       (client->cluster.state == MONGOC_CLUSTER_STATE_BORN)) {
+      mongoc_mutex_lock (&pool->mutex);
+      _mongoc_queue_push_tail (&pool->queue, client);
+   } else if (_mongoc_cluster_reconnect (&client->cluster, NULL)) {
+      mongoc_mutex_lock (&pool->mutex);
+      _mongoc_queue_push_tail (&pool->queue, client);
+   }else {
+      mongoc_client_destroy (client);
+      mongoc_mutex_lock (&pool->mutex);
+      pool->size--;
+   }
 
-   mongoc_mutex_lock(&pool->mutex);
-   _mongoc_queue_push_head(&pool->queue, client);
-   mongoc_cond_signal(&pool->cond);
-   mongoc_mutex_unlock(&pool->mutex);
+   mongoc_cond_signal (&pool->cond);
+   mongoc_mutex_unlock (&pool->mutex);
 
    EXIT;
 }
+
+uint
```

Comment:
I'd like to stick to explicitly sized types (out of stdint), or standard types.

So either returning size_t or uint32_t